ocaml/typing/typecore.ml

```diff
@@ -6903,10 +6903,14 @@ and type_cases
         let guard =
           match pc_guard with
           | None -> None
-          | Some scond ->
-              Some
-                (type_expect ext_env mode_local scond
-                   (mk_expected ~explanation:When_guard Predef.type_bool))
+          | Some scond -> (
+              match scond with
+              | Guard_predicate pred ->
+                Some
+                  (type_expect ext_env mode_local pred
+                    (mk_expected ~explanation:When_guard Predef.type_bool))
+              | _ -> (failwith "pattern guard typechecking unimplemented")
```

Comment:
We tend to avoid catch-all cases in `match` and instead explicitly write out the constructors. This guards against bugs where we add a constructor to the type and don't realize we need to update a match.